### DISCUSSION TITLE
k3s can't work well with coredns / kubedns , built by official website script

### DISCUSSION THREAD

**Discussion by 510mt** [k3s-io/k3s#3768] [Q&A]
**Environmental Info:**
K3s Version: 

k3s version v1.20.7+k3s1 
go version go1.15.12

Node(s) CPU architecture, OS, and Version: 

centos 7 
ubuntu 18/20

Cluster Configuration:

single server / 1 server 1 agent

**Describe the bug:**

after i installed k3s by the offical website script 
`curl -sfL https://get.k3s.io | sh -` 
i found that coredns is installed and ran well too ,
`root@mt510:/home/mt510# kubectl get pods -n kube-system -o wide|grep coredns

coredns-854c77959c-n5mpf                  1/1     Running     16         69d   10.42.0.152   mt510   <none>           <none>
`
but actually , after i deployed a service like nginx,  i can't access it by using curl (or nslookup / dig) nginx.nginx.svc.cluster.local ,
same to whatever pod i used, but i can nslookup / dig kubernetes.default.svc.cluster.local !!!
so i assume that coredns work well but some settings is wrong , maybe k3s or kubelet ? 

**Steps To Reproduce:**

- Installed K3s: 
curl -sfL https://get.k3s.io | sh -

**Expected behavior:**

create a nginx deployment and svc in nginx namespace, create annoter pod which with a curl / nslookup /dig util ,
and use the pod with the network utils to request the nginx by access nginx-xxxx.nginx.svc.cluster.local

**Actual behavior:**

it can't access the svc by request by the url nginx-xxxx.nginx.svc.cluster.local

**Additional context / logs:**

**Backporting**

- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
You are not expected to be able to resolve cluster addresses from outside containers running in the cluster. You'd normally expose things via LoadBalancer services, externaldns, etc.

**Reply by brandond** [reply]
Ah, my apologies, I had a hard time parsing your question - but the behavior you're looking for requires a service, not just a pod. Create a service named nginx in a namespace named nginx, and you can access it at nginx.nginx.svc.cluster.local

**Reply by brandond** [reply]
Ref: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/

**Reply by 510mt** [reply]
i created service for my pod, let me show you more details for this question, and i also fixed this bug by modify k3s.service  setting and /etc/resolve.conf , but i think it should be deal officially

Firstly k3s should be installed and runing well ,at the same time ,coredns runing well too. This step can process by the offical website install script. 
 
Then let's start with the application scenario
1.create namespace nginx
`kubectl create namespace nginx`

2.nginx deployment & service below
`
apiVersion: apps/v1
kind: Deployment
metadata:
 name: nginx
 namespace: nginx
spec:
 selector:
  matchLabels:
   app: web_server
 replicas: 1
 template:
  metadata:
   labels:
    app: web_server
  spec:
   containers:
   - name: nginx
     image: nginx:1.7.9
     ports:
     - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-app-svc
  namespace: nginx
spec:
  type: ClusterIP
  ports:
  - protocol: TCP
    port: 80
    targetPort: 80
  selector:
    app: web_server
`
2.check the svc
`kubectl get svc -n nginx`

3.create a cluster container to curl nginx by domain nginx-app-svc.nginx.svc.cluster.local
`
kubectl run curl --image=radial/busyboxplus:curl -i --tty --rm
`
then exec 'curl nginx-app-svc.nginx.svc.cluster.local' or use nslookup
but it can't response correctly (it should be work well by default because the kubelet helped ).

4.but i can set some arg to the k3s.service and node's /etc/resov.conf to fix the problem (actually i knew it can set by k8s dnsPolicy , but i think it should work by default)
`
kubectl get ep -n kube-system
`
to get the endPoint IP  , i got 10.42.0.5 here

then `/etc/systemd/system/k3s.service`
add below arg --kubelet-arg
`
ExecStart=/usr/local/bin/k3s \
    server \
        '--kubelet-arg=cluster-dns=10.42.0.5' \
`
finally set /etc/resolv.conf
`search default.svc.cluster.local svc.cluster.local cluster.local google.internal c.gce_project_id.internal
nameserver 10.42.0.5`

restart k3s ,then i can access by the cluster container and node directly 

**maybe i take the wrong way to make dns work correctly？？？**

**Reply by brandond** [reply]
It's not wrong per se, it's just  that cluster dns is not intended to be used outside of the cluster. Also, if you're going to do that, I would probably point your resolv.conf at the dns service, instead of at the pod address, as that will change every time the pod gets restarted. Also, don't set the cluster dns address to a pod address (outside the service cidr range) - that will break things.

**Reply by 510mt** [reply]
thank you!!!  i got that
